server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -4436,6 +4475,61 @@ private SnapshotsInProgress createSnapshot(
         }
     }
 
+    private Collection<LongWithAttributes> getSnapshotsInProgress() {
+        final ClusterState currentState = clusterService.state();
+        // Only the master should report on snapshots-in-progress
+        if (currentState.nodes().isLocalNodeElectedMaster() == false) {
+            return List.of();
+        }
+        final SnapshotsInProgress snapshotsInProgress = SnapshotsInProgress.get(currentState);
+        final List<LongWithAttributes> snapshotsInProgressMetrics = new ArrayList<>();
+        currentState.metadata().projects().forEach((projectId, project) -> {
+            final RepositoriesMetadata repositoriesMetadata = RepositoriesMetadata.get(project);
+            if (repositoriesMetadata != null) {
+                repositoriesMetadata.repositories().forEach(repository -> {
+                    int snapshotCount = snapshotsInProgress.forRepo(projectId, repository.name()).size();
+                    snapshotsInProgressMetrics.add(
+                        new LongWithAttributes(snapshotCount, SnapshotMetrics.createAttributesMap(projectId, repository))
+                    );
+                });
+            }
+        });
+        return snapshotsInProgressMetrics;
+    }
+
+    private Collection<LongWithAttributes> getShardsByState() {
+        final ClusterState currentState = clusterService.state();
+        // Only the master should report on shards-by-state
+        if (currentState.nodes().isLocalNodeElectedMaster() == false) {
+            return List.of();
+        }
+        final SnapshotsInProgress snapshotsInProgress = SnapshotsInProgress.get(currentState);
+        final List<LongWithAttributes> shardsByState = new ArrayList<>();
+        final ObjectIntMap<ShardState> shardCounts = new ObjectIntHashMap<>(TRACKED_SHARD_STATES.size());
+        currentState.metadata().projects().forEach((projectId, project) -> {
+            final RepositoriesMetadata repositoriesMetadata = RepositoriesMetadata.get(project);
+            if (repositoriesMetadata != null) {
+                for (RepositoryMetadata repository : repositoriesMetadata.repositories()) {
+                    TRACKED_SHARD_STATES.forEach(shardState -> shardCounts.put(shardState, 0));
+                    for (SnapshotsInProgress.Entry snapshot : snapshotsInProgress.forRepo(projectId, repository.name())) {
+                        for (ShardSnapshotStatus shardSnapshotStatus : snapshot.shards().values()) {
```

Comment:
I worry about these potentially-deeply-nested loops happening on each metric collection cycle. We're already keeping track of these things in `applyClusterState` (the only place they can change) - could we compute these counts there instead?

Reply:
Done in e90ff65f8bd2ac9c1d8f8a111a98e45a15bc01ec